### PR TITLE
Close on document click feature

### PR DESCRIPTION
**Pull request by daimoonis**
Hello,

Thank you for your work at fab-speed-dial library.

I would like to share my changes to my fork of your library. My changes are based on old material fab-speed-dial feature, that speed dial has been automatically closed on document click  [https://github.com/angular/material/blob/master/src/components/fabSpeedDial/fabController.js#L178](https://github.com/angular/material/blob/master/src/components/fabSpeedDial/fabController.js#L178)

What do you think about such feature as part of your library?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,6 +50,7 @@ or adapt the bindings to your needs:
 | `direction`        | `up`, `down`, `left` or `right` | `up`   | The direction to open the action buttons               |
 | `animationMode`    | `fling` or `scale` | `fling`             | The animation to apply when opening the action buttons |
 | `fixed`            | `boolean`          | `false`             | Indicates if this FAB Speed Dial is fixed (user cannot change the open state on click) |
+| `closeOnOutsideClick` | `boolean`       | `false`             | Indicates if this FAB Speed Dial will automatically close when user clicks somewhere in document (does not apply to fixed mode) |
```
Comment (PowerKiKi) [inline README.md]: I think this feature should be the only behavior possible and hence no configuration should be needed at all. This was the case in AngularJS AFAIK.

Is there a specific reason why you made it configurable ?

```suggestion
```

File: projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts [hunk 2]
```diff
@@ -103,12 +103,60 @@ describe('FabSpeedDial', () => {
...
+    it('should click on document testElement to hide all fabActions', () => {
+        const fixture = TestBed.createComponent(TestAppComponent);
+        const testComponent = fixture.debugElement.componentInstance;
+
+        const actionsSpy = spyOn(fixture.componentInstance.fabSpeedDial, 'setActionsVisibility').and.callThrough();
+        spyOn(fixture.componentInstance.fabActions, 'show').and.callThrough();
+        spyOn(fixture.componentInstance.fabActions, 'hide').and.callThrough();
+
+        testComponent.open = true;
+        testComponent.closeOnOutsideClick = true;
```
Comment (PowerKiKi) [inline projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts]: Without configuration, then this line should disappear

```suggestion
```

File: projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts [hunk 3]
```diff
@@ -103,12 +103,60 @@ describe('FabSpeedDial', () => {
...
+
+        const actionButton = fixture.debugElement.query(By.css('.testElement'));
+        actionButton.nativeElement.click();
+        fixture.detectChanges();
+
+        expect(fixture.componentInstance.fabSpeedDial.setActionsVisibility).toHaveBeenCalled();
+        expect(fixture.componentInstance.fabActions.hide).toHaveBeenCalled();
+    });
+
+    it('should click on document testElement but not hide fabActions', () => {
```
Comment (PowerKiKi) [inline projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts]: Without configuration, this entire test should be removed

File: projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts [hunk 4]
```diff
@@ -103,12 +103,60 @@ describe('FabSpeedDial', () => {
...
+    });
+
 });
 
 /** Test component that contains an fab speed dial buttons */
 @Component({
-    template: `
-        <eco-fab-speed-dial [direction]="direction" [(open)]="open" #fabSpeedDial>
+    template: `<div>
+        <eco-fab-speed-dial [direction]="direction" [(open)]="open" [closeOnOutsideClick]="closeOnOutsideClick" #fabSpeedDial>
```
Comment (PowerKiKi) [inline projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts]: Without configuration, this attribute would disappear too

```suggestion
        <eco-fab-speed-dial [direction]="direction" [(open)]="open" #fabSpeedDial>
```

File: projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts [hunk 5]
```diff
@@ -128,6 +178,7 @@ class TestAppComponent {
     public fabSpeedDial: EcoFabSpeedDialComponent;
     public direction = 'up';
     public open: boolean;
+    public closeOnOutsideClick: boolean;
```
Comment (PowerKiKi) [inline projects/fab-speed-dial/src/lib/fab-speed-dial.spec.ts]: Without configuration, then this line should disappear

```suggestion
```

File: projects/fab-speed-dial/src/lib/fab-speed-dial.ts [hunk 6]
```diff
@@ -182,18 +185,38 @@ export class EcoFabSpeedDialComponent implements AfterContentInit {
...
+        }
+        this._processOutsideClickState();
+    }
+
     @Output() openChange: EventEmitter<boolean> = new EventEmitter<boolean>();
 
     @ContentChild(EcoFabSpeedDialActionsComponent) _childActions: EcoFabSpeedDialActionsComponent;
 
-    constructor(private elementRef: ElementRef, private renderer: Renderer2) {
+    constructor(private elementRef: ElementRef, private renderer: Renderer2, @Inject(DOCUMENT) private document: any) {
```
Comment (PowerKiKi) [inline projects/fab-speed-dial/src/lib/fab-speed-dial.ts]: Should probably be typed better:

```suggestion
    constructor(private elementRef: ElementRef, private renderer: Renderer2, @Inject(DOCUMENT) private document: Document) {
```

File: projects/fab-speed-dial/src/lib/fab-speed-dial.ts [hunk 7]
```diff
@@ -111,12 +113,13 @@ export class EcoFabSpeedDialActionsComponent implements AfterContentInit {
...
 })
-export class EcoFabSpeedDialComponent implements AfterContentInit {
+export class EcoFabSpeedDialComponent implements OnDestroy, AfterContentInit {
     private isInitialized = false;
     private _direction: Direction = 'up';
     private _open = false;
     private _animationMode: AnimationMode = 'fling';
-
+    private _closeOnOutsideClick = false;
+    private _documentListener: () => void;
```
Comment (PowerKiKi) [inline projects/fab-speed-dial/src/lib/fab-speed-dial.ts]: This a function or null:

```suggestion
    private _documentListener: () => void | null = null;
```

File: projects/fab-speed-dial/src/lib/fab-speed-dial.ts [hunk 8]
```diff
@@ -230,6 +253,34 @@ export class EcoFabSpeedDialComponent implements AfterContentInit {
...
+                if (!this.fixed && this.open) {
+                    this.open = false;
+                }
+            });
+        }
+    }
+
+    private _unsetDocumentClickBindings() {
+        if (this._documentListener) {
+            this._documentListener();
```
Comment (PowerKiKi) [inline projects/fab-speed-dial/src/lib/fab-speed-dial.ts]: Probably missing a `this._documentListener = null` either here or in the listener itself... depends if we want to remove the listener after each click or not...

File: projects/fab-speed-dial/src/lib/fab-speed-dial.ts (hunk 6 above)
Comment (daimoonis) [reply projects/fab-speed-dial/src/lib/fab-speed-dial.ts]: When changing this i ran into [https://github.com/angular/angular/issues/20351](https://github.com/angular/angular/issues/20351)

Comment (daimoonis) [reply projects/fab-speed-dial/src/lib/fab-speed-dial.ts]: So i can set /** @dynamic */ comment before component and let 'Document' there or use 'any'.

File: README.md (hunk 1 above)
Comment (daimoonis) [reply README.md]: I will try to create it as the only behavior possible.

File: projects/fab-speed-dial/src/lib/fab-speed-dial.ts (hunk 8 above)
Comment (daimoonis) [reply projects/fab-speed-dial/src/lib/fab-speed-dial.ts]: Indeed it was missing. Anyway I also did change the listener registering/unregistering, it is registered when actions are visible and unregistered when hidden.

File: projects/fab-speed-dial/src/lib/fab-speed-dial.ts (hunk 6 above)
Comment (PowerKiKi) [reply projects/fab-speed-dial/src/lib/fab-speed-dial.ts]: I guess `@dynamic` would do the trick indeed

**Comment by PowerKiKi** [comment]
Thanks ! I made a few more minor changes (mostly naming) before merging.

**Comment by daimoonis** [comment]
Cool, are there any plans to release a version of your library with these changes?

**Comment by PowerKiKi** [comment]
Released 3.1.0

**Pull request closed without merging**